L12-csharp-interceptors.md

```diff
@@ -0,0 +1,212 @@
+
+An alternative design registering interceptors directly on a `Channel`
+returning an "intercepted `Channel`" object was considered but ultimately
+dismissed, despite having an advantage of 100% compatibility with anywhere a
+`Channel` would have been used, since it added complexity to the `Channel`
+object and potentially slowing down the fast code path where no interceptor is
+registered.  When invoking RPCs in user code, `CallInvoker` is sufficiently
+close to a `Channel` for constructing client objects and using local variable
+type inference (`var`), the actual user code would look identical in most
+cases:
+
+```csharp
+var interceptedChannel = channel.Intercept(interceptorObject);
+// interceptedChannel type is really `CallInvoker`, but since
+// both `CallInvoker` and `Channel` can be passed to the generated
+// client code, the code looks similar.
+var greeterClient = new GreeterClient(interceptedChannel);
+```
+
+A higher-level, more user-friendly, canonical base class for common interceptor
+functionality can be provided by deriving from `ClientInterceptor` and adding
+unified hooks (e.g. `BeginCall`, `EndCall`) that can be overridden once but
+operate on all types of RPC invocations.
+
+In order to carry satellite data across interceptors and `CallInvoker`s, this
+proposal suggests `CallOptions` to be amended with an `Items` dictionary and a
+`WithItems` helper method to register a custom dictionary which would hold
+shared state across the interceptors and `CallInvoker`s.  This is the only other
+piece of client interceptors that requires being implemented within the
+`Grpc.Core` since it needs adding a member to `CallOptions` structure.
+
+
+### Server Interceptors
+
+Server-side interceptors derive from the new `ServerInterceptor` class
+implemented in `Grpc.Core.Interceptors` namespace.  Server interceptors are
+registered on individual service definitions as opposed to an entire server,
+though it might be sensible to provide a syntactic sugar for adding an
+interceptor chain to all services served by a `Server` instance.
+
+In particular, an instance of a class derived from `ServerInterceptor` is
+registered via an extension method for class `ServerServiceDefinition`.  This
+method is defined as an extension method to decouple the server interceptor
+machinery from the core gRPC infrastructure as much as possible.  This
+extension method wraps all of the existing handlers in the service definition
+with ones intercepted by the passed interceptor object.
+
+```csharp
+Server server = new Server
+{
+	Services = { Greeter.BindService(new GreeterImpl()).Intercept(new LogInterceptor()) },
+	Ports = { new ServerPort("localhost", Port, ServerCredentials.Insecure) }
+};
+server.Start();
+```
+
+Since support for this handler substitution was needed from the
+`ServerServiceDefinition` class itself, a general internal method
+`SubstituteHandlers` is added to actually apply the mapping given the wrapping
```

Comment:
nit: if it's an internal method, it's probably irrelevant for the public API discussion here - it's purely a question of implementation.